File: libraries/libertscommunication/actuation_parameter_data.cpp
#include "actuation_parameter_data.h"

ActuationParameterData::ActuationParameterData(const uint8_t *data) {
    _ack_number = *(reinterpret_cast<const uint32_t *>(&data[0]));
    _rate_pitch_roll_lift = *(reinterpret_cast<const uint16_t *>(&data[4]));
    _rate_yaw = *(reinterpret_cast<const uint16_t *>(&data[6]));
    _motor_min = *(reinterpret_cast<const uint16_t *>(&data[8]));
    _motor_max = *(reinterpret_cast<const uint16_t *>(&data[10]));
}

ActuationParameterData::ActuationParameterData(uint16_t rate_roll_pitch_lift, uint16_t rate_yaw, uint16_t motor_min,
                                               uint16_t motor_max, uint32_t ack_number) {
    _ack_number = ack_number;
    _rate_pitch_roll_lift = rate_roll_pitch_lift;
    _rate_yaw = rate_yaw;
    _motor_min = motor_min;
    _motor_max = motor_max;
}

void ActuationParameterData::to_buffer(uint8_t *buffer) {
    *(reinterpret_cast<uint32_t *>(&buffer[0])) = _ack_number;
    *(reinterpret_cast<uint16_t *>(&buffer[4])) = _rate_pitch_roll_lift;
    *(reinterpret_cast<uint16_t *>(&buffer[6])) = _rate_yaw;
    *(reinterpret_cast<uint16_t *>(&buffer[8])) = _motor_min;
    *(reinterpret_cast<uint16_t *>(&buffer[10])) = _motor_max;
}

File: src/ERTS.Dashboard/Communication/Data/ActuationParameterData.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ERTS.Dashboard.Communication.Data
{
    class ActuationParameterData : PackageData
    {
        ushort ratePitchRollLift;
        ushort rateYaw;
        ushort motorMin;
        ushort motorMax;
        uint ackNumber;

        public ushort RatePitchRollLift {
            get { return ratePitchRollLift; }
        }
        public ushort RateYaw {
            get { return rateYaw; }
        }
        public ushort MotorMin {
            get { return motorMin; }
        }
        public ushort MotorMax {
            get { return motorMax; }
        }
        public uint AckNumber {
            get { return ackNumber; }
        }

        public ActuationParameterData(ushort RatePitchRollLift, ushort RateYaw, ushort MotorMin, ushort MotorMax, uint AckNumber)
        {
            ratePitchRollLift = RatePitchRollLift;
            rateYaw = RateYaw;
            motorMin = MotorMin;
            motorMax = MotorMax;
            ackNumber = AckNumber;
        }
        public ActuationParameterData(byte[] data)
        {
            if (data.Length < Length)
                throw new ArgumentException("Data is not long enough.", "data");

            ackNumber = BitConverter.ToUInt32(data, 0);
            ratePitchRollLift = BitConverter.ToUInt16(data, 4);
            rateYaw = BitConverter.ToUInt16(data, 6);
            motorMin = BitConverter.ToUInt16(data, 8);
            motorMax = BitConverter.ToUInt16(data, 10);
        }
        public ActuationParameterData()
        {
            throw new NotSupportedException();
        }

        public override int Length => 4 * sizeof(ushort) + sizeof(uint);

        public override bool ExpectsAcknowledgement => (true);

        public override bool IsValid()
        {
            return true;
        }

        public override byte[] ToByteArray()
        {
            byte[] data = new byte[Length];

            Buffer.BlockCopy(BitConverter.GetBytes(ackNumber), 0, data, 0, sizeof(uint));
            Buffer.BlockCopy(BitConverter.GetBytes(ratePitchRollLift), 0, data, 4, sizeof(ushort));
            Buffer.BlockCopy(BitConverter.GetBytes(rateYaw), 0, data, 6, sizeof(ushort));
            Buffer.BlockCopy(BitConverter.GetBytes(motorMin), 0, data, 8, sizeof(ushort));
            Buffer.BlockCopy(BitConverter.GetBytes(motorMax), 0, data, 10, sizeof(ushort));
            return data;
        }

        public override void SetAckNumber(uint AckNumber)
        {
            ackNumber = AckNumber;
        }

    }
}
